db/migrations/20230306004011_travel_overhaul.php

```diff
@@ -0,0 +1,208 @@
+               (2, 'Font of Vitality', 1, 10, 1, '/images/map/locations/fontofvitality.png', '38FF007E', 1, 1, 200),
+               (3, 'Stone', 1, 5, 3, '/images/village_icons/stone.png', '0000009f', 0, 0, 50),
+               (4, 'Cloud', 1, 17, 2, '/images/village_icons/cloud.png', '0000009f', 0, 0, 50),
+               (5, 'Leaf', 1, 9, 6, '/images/village_icons/leaf.png', '0000009f', 0, 0, 50),
+               (6, 'Sand', 1, 3, 8, '/images/village_icons/sand.png', '0000009f', 0, 0, 50),
+               (7, 'Mist', 1, 16, 10, '/images/village_icons/mist.png', '0000009f', 0, 0, 50);
+        ");
+        // maps_portal table
+        $this->execute("
+            CREATE TABLE `maps_portals`
+            (
+                `portal_id`     int(11)      NOT NULL AUTO_INCREMENT,
+                `from_id`       int(11)      NOT NULL,
+                `to_id`         int(11)      NOT NULL,
+                `entrance_x`    int(11)      NOT NULL,
+                `entrance_y`    int(11)      NOT NULL,
+                `entrance_name` varchar(255) NOT NULL,
+                `exit_x`        int(11)      NOT NULL,
+                `exit_y`        int(11)      NOT NULL,
+                `active`        BOOLEAN      NOT NULL DEFAULT 1,
+                `whitelist`     varchar(255) NOT NULL DEFAULT '[]',
+                PRIMARY KEY (`portal_id`)
+            ) ENGINE=MyISAM DEFAULT CHARSET=`latin1`;
+        ");
+        $this->execute("
+            INSERT INTO `maps_portals` 
+                (`portal_id`, `from_id`, `to_id`, `entrance_x`, `entrance_y`, `entrance_name`, `exit_x`,
+                            `exit_y`, `active`, `whitelist`)
+            VALUES (1, 1, 2, 5, 4, 'Forbidden', 1, 1, 1, 'Mist,Leaf,Stone,Sand,Cloud'),
+                   (2, 2, 1, 1, 1, 'Safety', 5, 4, 1, 'Mist,Leaf,Stone,Sand,Cloud');
+        ");
+        // users modifications
+
+        // add new columns
+        $this->execute("
+            ALTER TABLE `users`
+                ADD COLUMN `attack_id`          varchar(255) DEFAULT NULL,
+                ADD COLUMN `attack_id_time_ms`  bigint(14) DEFAULT NULL,
+                ADD COLUMN `filters`            TEXT DEFAULT NULL,
+                ADD COLUMN `last_movement_ms`   bigint(14) NOT NULL DEFAULT 0;
+        ");
+
+        // change from x.y to x:y:map_id
+        $stmt = $this->query("SELECT `user_id`, `location` FROM `users`");
+        while ($user_data = $stmt->fetch()) {
+            $loc_data = explode('.', $user_data['location']);
+            if (count($loc_data) > 1) {
+                $new_loc = $loc_data[0] . ':' . $loc_data[1];
+                $this->execute("UPDATE `users` SET `location`=? WHERE `user_id`=?",
+                    [$new_loc, $user_data['user_id']]
+                );
+            }
+        }
+
+        // add map_id to user location
+        $this->execute("
+            UPDATE `users`
+                SET `location` = concat(`location`, ':1');
+        ");
```

Comment:
The PHP way there totally works, but FYI there is a much simpler SQL alternative: https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_replace
```suggestion
  $this->execute("UPDATE `users` SET `location`=
    CONCAT(
      REPLACE(`location`, '.', ':'), 
      ':1'
    )
  ");
```

Reply:
And for `down()` you can use 
```mysql
UPDATE users SET location = REPLACE(
  SUBSTRING_INDEX(`location`, ":", 2), 
  ":", 
  "."
)
```

Reply:
`SUBSTRING_INDEX()` (nifty function, TIL!) will return the substring up until `count` instances of `delim`, which makes it perfect for returning the first two parts of a location like`20:2:1`

Reply:
Bruh, forreals. It's like 3 lines.